File: Top200/2-AddTwoNumbers.java
package Top200;

class RemoveNthFromEnd {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1);
        dummy.next = null;
        ListNode curr = dummy;
        int carry = 0;

        while (l1 != null && l2 != null) {
            int sum = l1.val + l2.val + carry;
            carry = sum / 10;
            sum = sum % 10;

            curr.next = new ListNode(sum, null);
            
            curr = curr.next;
            l1 = l1.next;
            l2 = l2.next;
        }

        while (l1 != null) {
            int sum = l1.val + carry;
            carry = sum / 10;
            sum = sum % 10;

            curr.next = new ListNode(sum, null);

            curr = curr.next;
            l1 = l1.next;
        }

        while (l2 != null) {
            int sum = l2.val + carry;
            carry = sum / 10;
            sum = sum % 10;

            curr.next = new ListNode(sum, null);

            curr = curr.next;
            l2 = l2.next;
        }

        if (carry != 0) {
            curr.next = new ListNode(carry, null);
        }

        return dummy.next;
    }   
}

File: Top200/2-AddTwoNumbers.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{Val: -1}
	curr := dummy
	curr.Next = nil
	carry := 0

	for l1 != nil && l2 != nil {
		sum := l1.Val + l2.Val + carry
		carry = sum / 10
		sum = sum % 10

		curr.Next = &ListNode{Val: sum, Next: nil}
		curr = curr.Next

		l1 = l1.Next
		l2 = l2.Next
	}

	for l1 != nil {
		sum := l1.Val + carry
		carry = sum / 10
		sum = sum % 10

		curr.Next = &ListNode{Val: sum, Next: nil}
		curr = curr.Next

		l1 = l1.Next
	}

	for l2 != nil {
		sum := l2.Val + carry
		carry = sum / 10
		sum = sum % 10

		curr.Next = &ListNode{Val: sum, Next: nil}
		curr = curr.Next

		l2 = l2.Next
	}

	if carry != 0 {
		curr.Next = &ListNode{Val: carry, Next: nil}
	}

	return dummy.Next
}
